File: ordo-editor/packages/core/src/model/step.ts
/**
 * Step types for the Ordo rule engine
 * 步骤类型定义
 */

import { Condition } from './condition';
import { Expr } from './expr';

/** Step types */
export type StepType = 'decision' | 'action' | 'terminal' | 'sub_rule';

/** Base step interface */
export interface BaseStep {
  /** Unique step ID */
  id: string;
  /** Display name */
  name: string;
  /** Optional description */
  description?: string;
  /** Step type */
  type: StepType;
  /** Position in the flow editor (for visualization) */
  position?: {
    x: number;
    y: number;
  };
  /** Internal runtime marker; generated steps should not surface as authoring suggestions. */
  systemGenerated?: 'sub_rule_runtime' | 'concept_runtime';
}

/** Branch definition for decision steps */
export interface Branch {
  /** Unique branch ID */
  id: string;
  /** Branch label (for display) */
  label?: string;
  /** Condition to evaluate */
  condition: Condition;
  /** Next step ID if condition is true */
  nextStepId: string;
}

/** Decision step - evaluates conditions and branches */
export interface DecisionStep extends BaseStep {
  type: 'decision';
  /** Ordered list of branches (evaluated in order) */
  branches: Branch[];
  /** Default next step if no branch matches */
  defaultNextStepId: string;
}

/** Variable assignment */
export interface VariableAssignment {
  /** Variable name (without $ prefix) */
  name: string;
  /** Value expression */
  value: Expr;
}

/** External call configuration */
export interface ExternalCall {
  /** Call type */
  type: 'http' | 'grpc' | 'function';
  /** Target URL or function name */
  target: string;
  /** Request parameters */
  params?: Record<string, Expr>;
  /** Variable to store result */
  resultVariable?: string;
  /** Timeout in milliseconds */
  timeout?: number;
  /** Retry configuration */
  retry?: {
    maxAttempts: number;
    backoffMs: number;
  };
  /** Error handling */
  onError?: 'fail' | 'ignore' | 'fallback';
  /** Fallback value on error */
  fallbackValue?: Expr;
}

/** Action step - performs operations */
export interface ActionStep extends BaseStep {
  type: 'action';
  /** Variable assignments */
  assignments?: VariableAssignment[];
  /** External calls */
  externalCalls?: ExternalCall[];
  /** Logging configuration */
  logging?: {
    message: Expr;
    level?: 'debug' | 'info' | 'warn' | 'error';
  };
  /** Next step ID */
  nextStepId: string;
}

/** Output field definition */
export interface OutputField {
  /** Field name */
  name: string;
  /** Value expression */
  value: Expr;
}

/** Terminal step - ends execution with a result */
export interface TerminalStep extends BaseStep {
  type: 'terminal';
  /** Result code (e.g., "APPROVED", "REJECTED") */
  code: string;
  /** Result message (can be expression) */
  message?: Expr;
  /** Output fields */
  output?: OutputField[];
}

/** Input binding for a sub-rule: inject a parent-context expression into a named input field */
export interface SubRuleBinding {
  /** Field name in the child context */
  field: string;
  /** Expression evaluated in the parent context */
  expr: Expr;
}

/** Output mapping for a sub-rule: copy a child variable back to a parent variable */
export interface SubRuleOutput {
  /** Variable name in the parent context (without $ prefix) */
  parentVar: string;
  /** Variable name in the child context (without $ prefix) */
  childVar: string;
}

/** Managed SubRule asset reference. Sub-rules are snapshotted inline when the parent is published. */
export interface SubRuleAssetRef {
  scope: 'project' | 'org';
  name: string;
}

/** Sub-rule step - executes a managed SubRule asset and returns control to the parent */
export interface SubRuleStep extends BaseStep {
  type: 'sub_rule';
  /** Legacy/engine-compatible reference name. Prefer assetRef for Studio assets. */
  refName: string;
  /** Managed SubRule asset reference used by Studio and platform resolution. */
  assetRef?: SubRuleAssetRef;
  /** Input bindings: expressions from the parent context injected into the child context */
  bindings?: SubRuleBinding[];
  /** Output mappings: variables from the child context written back to the parent context */
  outputs?: SubRuleOutput[];
  /** Authoring-level return semantics. Execution lowering materializes runtime control flow. */
  returnPolicy?: 'continue' | 'propagate_terminal';
  /** Next step ID after the sub-rule completes. Empty means the child terminal result ends the parent flow. */
  nextStepId: string;
}

/** Step union type */
export type StepUnion = DecisionStep | ActionStep | TerminalStep | SubRuleStep;

// ============================================================================
// Step builder helpers
// ============================================================================

let stepIdCounter = 0;

function generateStepId(): string {
  return `step_${++stepIdCounter}`;
}

function generateBranchId(): string {
  return `branch_${++stepIdCounter}`;
}

/** Alias for Step type (backward compatibility) */
export type Step = StepUnion;

export const Step = {
  /** Reset ID counter (for testing) */
  resetIdCounter() {
    stepIdCounter = 0;
  },

  /** Create a decision step */
  decision(options: {
    id?: string;
    name: string;
    description?: string;
    branches?: Branch[];
    defaultNextStepId: string;
    position?: { x: number; y: number };
    systemGenerated?: BaseStep['systemGenerated'];
  }): DecisionStep {
    return {
      id: options.id || generateStepId(),
      name: options.name,
      description: options.description,
      type: 'decision',
      branches: options.branches || [],
      defaultNextStepId: options.defaultNextStepId,
      position: options.position,
      systemGenerated: options.systemGenerated,
    };
  },

  /** Create an action step */
  action(options: {
    id?: string;
    name: string;
    description?: string;
    assignments?: VariableAssignment[];
    externalCalls?: ExternalCall[];
    logging?: ActionStep['logging'];
    nextStepId: string;
    position?: { x: number; y: number };
    systemGenerated?: BaseStep['systemGenerated'];
  }): ActionStep {
    return {
      id: options.id || generateStepId(),
      name: options.name,
      description: options.description,
      type: 'action',
      assignments: options.assignments,
      externalCalls: options.externalCalls,
      logging: options.logging,
      nextStepId: options.nextStepId,
      position: options.position,
      systemGenerated: options.systemGenerated,
    };
  },

  /** Create a terminal step */
  terminal(options: {
    id?: string;
    name: string;
    description?: string;
    code: string;
    message?: Expr;
    output?: OutputField[];
    position?: { x: number; y: number };
    systemGenerated?: BaseStep['systemGenerated'];
  }): TerminalStep {
    return {
      id: options.id || generateStepId(),
      name: options.name,
      description: options.description,
      type: 'terminal',
      code: options.code,
      message: options.message,
      output: options.output,
      position: options.position,
      systemGenerated: options.systemGenerated,
    };
  },

  /** Create a branch */
  branch(options: {
    id?: string;
    label?: string;
    condition: Condition;
    nextStepId: string;
  }): Branch {
    return {
      id: options.id || generateBranchId(),
      label: options.label,
      condition: options.condition,
      nextStepId: options.nextStepId,
    };
  },

  /** Create a sub-rule step */
  subRule(options: {
    id?: string;
    name: string;
    description?: string;
    refName: string;
    assetRef?: SubRuleAssetRef;
    bindings?: SubRuleBinding[];
    outputs?: SubRuleOutput[];
    returnPolicy?: SubRuleStep['returnPolicy'];
    nextStepId: string;
    position?: { x: number; y: number };
    systemGenerated?: BaseStep['systemGenerated'];
  }): SubRuleStep {
    return {
      id: options.id || generateStepId(),
      name: options.name,
      description: options.description,
      type: 'sub_rule',
      refName: options.refName,
      assetRef: options.assetRef,
      bindings: options.bindings,
      outputs: options.outputs,
      returnPolicy: options.nextStepId ? options.returnPolicy ?? 'continue' : 'propagate_terminal',
      nextStepId: options.nextStepId,
      position: options.position,
      systemGenerated: options.systemGenerated,
    };
  },

  /** Create a variable assignment */
  assign(name: string, value: Expr): VariableAssignment {
    return { name, value };
  },

  /** Create an output field */
  output(name: string, value: Expr): OutputField {
    return { name, value };
  },
};

/** Check if a step is a decision step */
export function isDecisionStep(step: Step): step is DecisionStep {
  return step.type === 'decision';
}

/** Check if a step is an action step */
export function isActionStep(step: Step): step is ActionStep {
  return step.type === 'action';
}

/** Check if a step is a terminal step */
export function isTerminalStep(step: Step): step is TerminalStep {
  return step.type === 'terminal';
}

/** Check if a step is a sub-rule step */
export function isSubRuleStep(step: Step): step is SubRuleStep {
  return step.type === 'sub_rule';
}

/** Get the next step IDs from a step */
export function getNextStepIds(step: Step): string[] {
  switch (step.type) {
    case 'decision':
      return [...step.branches.map((b) => b.nextStepId), step.defaultNextStepId].filter(
        (id, i, arr) => arr.indexOf(id) === i
      ); // unique

    case 'action':
      return [step.nextStepId];

    case 'terminal':
      return [];

    case 'sub_rule':
      return step.nextStepId ? [step.nextStepId] : [];
  }
}

File: crates/ordo-protocol/src/types/step.rs
//! Studio step types (mirrors the TypeScript Step model)

use serde::{Deserialize, Serialize};

use super::condition::StudioCondition;
use super::expr::StudioExpr;

/// A step in the studio format.
///
/// The `type` field is flattened from `StudioStepKind` and discriminates the step kind.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioStep {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    // position is ignored during conversion (visual-only)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub position: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub system_generated: Option<String>,
    #[serde(flatten)]
    pub kind: StudioStepKind,
}

/// Discriminated step kind — `type` field in JSON.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StudioStepKind {
    Decision {
        #[serde(default)]
        branches: Vec<StudioBranch>,
        #[serde(rename = "defaultNextStepId", default)]
        default_next_step_id: Option<String>,
    },
    Action {
        #[serde(default)]
        assignments: Vec<StudioAssignment>,
        #[serde(rename = "externalCalls", default)]
        external_calls: Vec<StudioExternalCall>,
        #[serde(default)]
        logging: Option<StudioLogging>,
        #[serde(rename = "nextStepId")]
        next_step_id: String,
    },
    Terminal {
        code: String,
        #[serde(default)]
        message: Option<StudioTerminalMessage>,
        #[serde(default)]
        output: Vec<StudioOutputField>,
    },
    #[serde(rename = "sub_rule")]
    SubRule {
        #[serde(rename = "refName")]
        ref_name: String,
        #[serde(default)]
        bindings: Vec<StudioSubRuleBinding>,
        #[serde(default)]
        outputs: Vec<StudioSubRuleOutput>,
        #[serde(rename = "returnPolicy", default)]
        return_policy: Option<String>,
        #[serde(rename = "nextStepId")]
        next_step_id: String,
    },
}

/// Terminal message accepts both the modern expression object shape and the
/// legacy plain string shape for backward compatibility.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum StudioTerminalMessage {
    Expr(StudioExpr),
    String(String),
}

/// A branch in a decision step.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioBranch {
    pub id: String,
    #[serde(default)]
    pub label: Option<String>,
    pub condition: StudioCondition,
    #[serde(rename = "nextStepId")]
    pub next_step_id: String,
}

/// A variable assignment in an action step.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioAssignment {
    pub name: String,
    pub value: StudioExpr,
}

/// An external service call in an action step.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StudioExternalCall {
    #[serde(rename = "type")]
    pub call_type: String,
    pub target: String,
    #[serde(default)]
    pub params: std::collections::HashMap<String, StudioExpr>,
    #[serde(default)]
    pub result_variable: Option<String>,
    #[serde(default)]
    pub timeout: Option<u64>,
}

/// Logging configuration in an action step.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioLogging {
    pub message: StudioExpr,
    #[serde(default)]
    pub level: Option<String>,
}

/// An output field in a terminal step.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioOutputField {
    pub name: String,
    pub value: StudioExpr,
}

/// Input binding for a sub-rule step.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioSubRuleBinding {
    pub field: String,
    pub expr: StudioExpr,
}

/// Output mapping for a sub-rule step.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StudioSubRuleOutput {
    pub parent_var: String,
    pub child_var: String,
}
